rfcs/convergence/make-overrides.md

```diff
@@ -0,0 +1,182 @@
+- Try to write a neutral problem statement, not one that motivates your desired solution
+- Remember, "Writing is thinking". It's natural to realize new ideas while writing your proposal
+-->
+
+---
+
+@layershifter
+
+## Summary
+
+This RFC proposes to replace existing `makeStyles()` implementation with a new one, i.e. proposes API changes.
+
+## Background
+
+In [microsoft/fluentui#16082](https://github.com/microsoft/fluentui/pull/16082) we have implemented a version of `makeStyles()` that uses atomic classes. Recent work on `makeOverrides()` shows significant performance improvements [microsoft/fluentui#16923](https://github.com/microsoft/fluentui/pull/16923) (around 15%) by changes in the implementation design.
+
+I will use `makeOverrides()` and `makeStyles()` for naming in this RFC to avoid confusion, but a proposal is to replace the current `makeStyles()` implementation with `makeOverrides()`.
+
+## Problem statement
+
+`makeStyles()` implementation has two problems that will cause issues at scale: matchers and missing slots. A snippet below will be used to highlight issues:
+
+```tsx
+import { ax, makeStyles } from "@fluentui/react-make-styles";
+
+const useRootStyles = makeStyles<TSelectors>([
+  [null, { color: "red" }],
+  [
+    s => s.primary /* <- a matcher function, will be executed on each render */,
+    { color: "blue" }
+  ]
+]);
+
+const useIconStyles = makeStyles<TSelectors>([
+  /* styles for each slot are defined separately */
+  [null, { background: "black" }],
+  [s => s.primary, { background: "white" }]
+]);
+
+function Component() {
+  const rootClasses = useRootStyles();
+
+  // React hooks cannot be called conditionally, https://reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level
+  // so we *have to* render styles even if they are not used in markup
+  const iconClasses = useIconStyles();
+
+  return (
+    <div className={ax(rootClasses, props.className)}>
+      {props.icon && <div className={ax(iconClasses, props.icon.className)} />}
+    </div>
+  );
+}
+```
+
+To compute classnames we need to understand what styles should be applied and thus execute matchers on each render as memoization of user's input is more expensive.
+
+This moves us to a next issue: there is no way to define styles for multiple slots/components at the same time in API. In this case, we will need to have a separate call of `makeStyles()` for each slot and because these calls are represented by React hooks we can't bailout even if these slots will not be rendered.
+
+_Side note:_ Initially we have been focused on implementation of atomic CSS and merging style definitions. As a source for inspirations we have used [Facebook stylex talk](https://www.youtube.com/watch?v=9JZHodNR184):
+
+- there was nothing like [`ax()`](https://github.com/microsoft/fluentui/pull/16411) function to merge atomic classes.
+- there was no contextual RTL support (parts of app can be rendered with different text directions)
```

Comment:
I don't quite understand, are these relted to the side note above ? Perhaps put all this in the same quote block ?

Reply:
Played with formatting in this block, is it cleaner?

Reply:
yeah a lot better IMO